reference-implementation/lib/readable-stream.js

```
@@ -86,273 +89,126 @@ class ReadableStream {
-        // closed before pipeTo(). It's possible but fragile.
-        _writer.close().then(
-          () => {
-            return _allWrites;
-          },
-          reason => {
-            releaseWriter();
-            finishWithRejection(reason);
-          }
-        ).then(
-          () => {
-            releaseWriter();
-            finishWithFulfillment();
-          }
-        );
-        _state = 'closingDest';
+      // Closing must be propagated backward
+      writer._closedPromise.then(() => {
+        const destClosed = new TypeError('the destination writable stream closed before all data could be piped to it');
 
-        return;
-      }
+        if (preventCancel === false) {
+          performAsyncShutdown(() => ReadableStreamCancel(this, destClosed), true, destClosed);
+        } else {
+          performShutdown(true, destClosed);
+        }
+      });
 
-      if (_lastWrite === undefined) {
-        releaseWriter();
-        finishWithFulfillment();
-        return;
+      function waitForCurrentWrite() {
+        return currentWrite.catch(() => {});
       }
 
-      // We don't use writer.closed. pipeTo() is responsible only for what it has written.
-      _lastWrite.then(
-        () => {
-          releaseWriter();
-          finishWithFulfillment();
-        },
-        reason => {
-          releaseWriter();
-          finishWithRejection(reason);
+      function performAsyncShutdown(action, originalIsError, originalError) {
+        if (shuttingDown === true) {
+          return;
         }
-      );
-      _state = 'waitingLastWriteOnReadableClosed';
-    }
-
-    function doPipe() {
-      // console.log('pipeTo(): doPipe()');
+        shuttingDown = true;
+
+        waitForCurrentWrite().then(() => {
+          action().then(
+            () => performShutdown(originalIsError, originalError),
+            newError => performShutdown(true, newError)
```

It's nice to propagate all the errors to someone, but it means that when preventCancel (or preventAbort) is not in use, the original cause of error is lost. It sounds bad.